### PR TITLE
Add support for 'force_auth'

### LINKED ISSUE #1196: [fxa-client] add support for 'force_auth'
see https://github.com/mozilla-mobile/fenix/issues/2833

┆Issue is synchronized with this [Jira Story](https://jira.allizom.org/browse/SYNC-31)
┆Story Points: 5
┆Sprint: SYNC - 2019-07-19
┆Epic: Fenix 2.0 Support

### PR DESCRIPTION
**Pull request by vladikoff**
Fixes #1196

### Pull Request checklist ###

- [x] **Quality**: This PR builds and tests run cleanly
  - `cargo test --all` produces no test failures
  - `cargo clippy --all --all-targets --all-features` runs without emitting any warnings
  - `cargo fmt` does not produce any changes to the code
  - `./gradlew ktlint detekt` runs without emitting any warnings
  - `swiftformat --swiftversion 4 megazords components/*/ios && swiftlint` runs without emitting any warnings or producing changes
  - Note: For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes a changelog entry or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [x] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/master/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

### CODE REVIEW

File: components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FirefoxAccount.kt [hunk 1]
```diff
@@ -103,6 +103,22 @@ class FirefoxAccount(handle: FxaHandle, persistCallback: PersistCallback?) : Aut
...
+    /**
+     * Constructs a URL used to begin the `force_auth` OAuth flow for the requested email, scopes and keys.
+     *
+     * This performs network requests, and should not be used on the main thread.
+     *
+     * @param scopes List of OAuth scopes for which the client wants access
+     * @param wantsKeys Fetch keys for end-to-end encryption of data from Mozilla-hosted services
+     * @return String that resolves to the flow URL when complete
+     */
+    fun beginForceAuthOAuthFlow(email: String, scopes: Array<String>, wantsKeys: Boolean): String {
```
Comment (eoger) [inline components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FirefoxAccount.kt]: Why not re-use `beginOAuthFlow` with an optional email param?

Comment (vladikoff) [reply components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FirefoxAccount.kt]: good plan! In Kt and Swift only?

Comment (eoger) [reply components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FirefoxAccount.kt]: Besides the URL params how different is the rust code?

Comment (vladikoff) [reply components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FirefoxAccount.kt]: 1 param is different (action), 1 param is new (email)

Comment (rfk) [reply components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FirefoxAccount.kt]: FWIW, I can imagine a world in which the caller might want to provide an email address as a hint (say, one they captured in an about:welcome style promotion during first-run) but don't care the enforce that the user *must* authenticate as that email address.

File: components/fxa-client/src/oauth.rs [hunk 2]
```diff
@@ -122,6 +122,40 @@ impl FirefoxAccount {
...
+                all_scopes.extend(existing_scopes);
+                HashSet::<String>::from_iter(all_scopes)
+                    .into_iter()
+                    .collect()
+            }
+            None => scopes.iter().map(ToString::to_string).collect(),
+        };
+        let scopes: Vec<&str> = scopes.iter().map(<_>::as_ref).collect();
+        self.oauth_flow(url, &scopes, wants_keys)
+    }
```
Comment (rfk) [inline components/fxa-client/src/oauth.rs]: One improvement we could make here, would be to explicitly tag the flow as a "force-auth" flow in self.flow_store. Then `complete_oauth_flow` can double-check whether the user did indeed re-auth as the intended user account, which might help us detect or avoid some unexpected edge-cases.

**Comment by vladikoff** [comment]
Thank you for the comments @rfk and @eoger 👍

**Comment by vladikoff** [comment]
Closing this for now. We shall use the strategy from https://github.com/mozilla/application-services/pull/1378

**Pull request closed without merging**